Scaffoldingllm supports MCP

wu1du2 opened the pull request:
# Support MCP in TensorRT-LLM
Scaffolding support MCP #3335 https://github.com/NVIDIA/TensorRT-LLM/issues/3335

## Description
MCP provides a standard tool-use ability to TensorRT-LLM, to utilize LLM function-call.

## examples
run a mcp server first. TRT-LLM connects to it, then processing function-call automatically.
see examples\scaffolding\contrib\mcp\README.md

WeiHaocheng commented on tensorrt_llm/scaffolding/contrib/mcp/mcp_controller.py:
```diff
@@ -0,0 +1,80 @@
+    class WorkerTag(Enum):
+        GENERATION = "generation"
+        MCPLIST = "mcp"
+        MCPCALL = "mcp"
```
MCPLIST and MCPCALL can be merged as a single tag: "MCP"

WeiHaocheng commented on tensorrt_llm/scaffolding/contrib/mcp/mcp_worker.py:
```diff
@@ -0,0 +1,39 @@
+        return TaskStatus.SUCCESS
+
+    async def shutdown(self):
+        await self.mcp_client.cleanup()
```
"shutdown" of worker is not a "async def"

WeiHaocheng replied:
You should run a asyncio task in a sync def. Likes this:
<img width="557" alt="image" src="https://github.com/user-attachments/assets/24bb6dfd-e33c-4da7-a529-ccfe649af67f" />

WeiHaocheng commented on tensorrt_llm/scaffolding/worker.py:
```diff
@@ -1,6 +1,5 @@
 from abc import ABC
-from copy import deepcopy
-from typing import Callable, List, Optional, Union
+from typing import Callable, Optional, Union, List
```
I guess the order should not be changed with running "pre-commit run --all-files".

WeiHaocheng commented on tensorrt_llm/scaffolding/contrib/mcp/chat_task.py:
```diff
@@ -0,0 +1,16 @@
+class ChatTask(GenerationTask):
+    messages : list = None
+    tools = None
+    choice = None
```
We can store more specific things here, "choice" is too broad

VALLIS-NERIA commented on examples/scaffolding/contrib/mcp/README.md:
```diff
@@ -0,0 +1,12 @@
+USAGE
```
Please use Markdown syntax

VALLIS-NERIA commented on examples/scaffolding/contrib/mcp/mcptest.py:
```diff
@@ -0,0 +1,46 @@
+    return 
+
+if __name__ == '__main__':
+    asyncio.run(main())
```
Please install pre-commit to format source code automatically

VALLIS-NERIA commented on tensorrt_llm/scaffolding/contrib/__init__.py:
```diff
@@ -1,12 +1,20 @@
 
-from .AsyncGeneration import StreamGenerationTask, stream_generation_handler
-from .Dynasor import DynasorGenerationController
+# from .AsyncGeneration import StreamGenerationTask, stream_generation_handler
```
`StreamGenerationTask` is still used by other code. Do you forget to revert the test purpose change?

wu1du2 committed:
remove pics

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 commented:
I’ve made the changes as recommended above, but calling MCPClient.shutdown may still trigger a GeneratorExit error, and I haven’t been able to resolve it yet.

WeiHaocheng commented:
/bot run

tensorrt-cicd commented:
[PR_Github #5974](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5974/) [ run ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #5974](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/5974/) [ run ] completed with state `FAILURE`
[/LLM/main/L0_MergeRequest_PR pipeline #4368](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4368/) completed with status: 'FAILURE'

wu1du2 committed:
fix spell

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 commented:
I've made some changes to pass the pre-commit. Please review the code and run the CI bot again.

WeiHaocheng commented:
/bot run

tensorrt-cicd commented:
[PR_Github #6069](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6069/) [ run ] triggered by Bot

wu1du2 committed:
support mcp

# Conflicts:
#	tensorrt_llm/scaffolding/worker.py

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 committed:
move all into contrib/mcp

# Conflicts:
#	examples/scaffolding/contrib/mcp/mcptest.py
#	tensorrt_llm/scaffolding/__init__.py
#	tensorrt_llm/scaffolding/contrib/__init__.py
#	tensorrt_llm/scaffolding/contrib/mcp/__init__.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_controller.py
#	tensorrt_llm/scaffolding/task.py
#	tensorrt_llm/scaffolding/worker.py

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 committed:
support sandbox, websearch

# Conflicts:
#	examples/scaffolding/contrib/mcp/mcptest.py
#	examples/scaffolding/contrib/mcp/weather/weather.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_controller.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_utils.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_worker.py
#	tensorrt_llm/scaffolding/worker.py

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 committed:
pre-commit fix

# Conflicts:
#	tensorrt_llm/scaffolding/contrib/mcp/__init__.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_utils.py
#	tensorrt_llm/scaffolding/contrib/mcp/mcp_worker.py

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 committed:
rebase

Signed-off-by: wu1du2 <wu1du2@gmail.com>

wu1du2 commented:
Rebase to current main branch.

tensorrt-cicd commented:
[PR_Github #6069](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6069/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4440](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4440/) completed with status: 'FAILURE'

WeiHaocheng commented:
/bot run

tensorrt-cicd commented:
[PR_Github #6109](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6109/) [ run ] triggered by Bot

WeiHaocheng commented:
/bot run

tensorrt-cicd commented:
[PR_Github #6109](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6109/) [ run ] completed with state `SUCCESS`
[/LLM/main/L0_MergeRequest_PR pipeline #4464](https://nv/trt-llm-cicd/job/main/job/L0_MergeRequest_PR/4464/) completed with status: 'SUCCESS'

WeiHaocheng commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #6195](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6195/) [ reuse-pipeline ] triggered by Bot

WeiHaocheng commented:
/bot reuse-pipeline

tensorrt-cicd commented:
[PR_Github #6196](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6196/) [ reuse-pipeline ] triggered by Bot

tensorrt-cicd commented:
[PR_Github #6196](https://nv/trt-llm-cicd/job/helpers/job/PR_Github/6196/) [ reuse-pipeline ] completed with state `SUCCESS`
Reusing [PR_Github #6109](https://prod.blsm.nvidia.com/sw-tensorrt-top-1/job/LLM/job/helpers/job/PR_Github/6109/) for commit `f74e9f3`
